File: Go/buy_stock.go
func maxProfit(prices []int) int {
    if len(prices) < 2 {return 0}
    profit := 0
    buy := prices[0]

    for i := 1; i < len(prices); i++ {
        if prices[i] < buy {
            buy = prices[i]
        } else {
            profit = max(profit, prices[i] - buy)
        }
    }
    return profit
}

func max(a, b int) int {
    if a > b {return a}
    return b
}

File: Java/BuyStock.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices == null || prices.length < 2) return 0;
        int profit = 0, buy = prices[0];

        for(int i = 1; i < prices.length; i++) {
            if(prices[i] < buy) buy = prices[i];
            else profit = Math.max(profit, prices[i] - buy);
        }

        return profit;
    }
}